File: apps/api/src/routes/domains.ts
import { randomUUID } from 'crypto'
import type { FastifyInstance } from 'fastify'
import type {
  CreateDomainInput,
  VerifyDomainInput,
  Domain,
  DomainVerificationInstructions,
} from '@recon/core'
import { query, queryOne, transaction } from '../lib/db.js'
import { authenticate } from '../middleware/authenticate.js'
import { verifyDnsTxt, verifyWellKnownFile } from '../services/domain-verification.js'

// Accepts apex domains and subdomains: example.com, api.example.com, etc.
const DOMAIN_RE = /^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}$/i

function buildInstructions(
  domain: string,
  token: string,
): DomainVerificationInstructions {
  return {
    domain,
    token,
    dns_txt: {
      record_name: `_recon-verify.${domain}`,
      record_type: 'TXT',
      record_value: token,
    },
    well_known_file: {
      url: `https://${domain}/.well-known/recon-verification.txt`,
      file_path: '.well-known/recon-verification.txt',
      content: token,
    },
  }
}

export async function domainRoutes(app: FastifyInstance) {
  // All domain routes require auth
  app.addHook('preHandler', authenticate)

  // GET /api/v1/domains
  app.get('/domains', async (req, reply) => {
    const domains = await query<Domain>(
      `SELECT id, user_id, domain, verification_status, verification_method,
              verification_token, verified_at, created_at, updated_at
       FROM domains
       WHERE user_id = $1
       ORDER BY created_at DESC`,
      [req.user.sub],
    )
    return reply.send({ domains })
  })

  // POST /api/v1/domains
  app.post<{ Body: CreateDomainInput }>('/domains', async (req, reply) => {
    const { domain } = req.body ?? {}

    if (!domain || !DOMAIN_RE.test(domain)) {
      return reply.status(400).send({ error: 'Valid domain name required (e.g. example.com)' })
    }

    const normalized = domain.toLowerCase().trim()

    const existing = await queryOne(
      'SELECT id FROM domains WHERE user_id = $1 AND domain = $2',
      [req.user.sub, normalized],
    )
    if (existing) {
      return reply.status(409).send({ error: 'Domain already registered for your account' })
    }

    const token = `recon-verify-${randomUUID()}`

    const row = await queryOne<Domain>(
      `INSERT INTO domains (user_id, domain, verification_token)
       VALUES ($1, $2, $3)
       RETURNING *`,
      [req.user.sub, normalized, token],
    )
    if (!row) throw new Error('Domain insert failed')

    // Audit
    await query(
      `INSERT INTO audit_log (user_id, action, target, ip_address, metadata)
       VALUES ($1, 'domain_registered', $2, $3::inet, $4)`,
      [
        req.user.sub,
        normalized,
        req.ip || null,
        JSON.stringify({ domain_id: row.id }),
      ],
    )

    return reply.status(201).send({
      domain: row,
      instructions: buildInstructions(normalized, token),
    })
  })

  // GET /api/v1/domains/:id
  app.get<{ Params: { id: string } }>('/domains/:id', async (req, reply) => {
    const row = await queryOne<Domain>(
      'SELECT * FROM domains WHERE id = $1 AND user_id = $2',
      [req.params.id, req.user.sub],
    )
    if (!row) return reply.status(404).send({ error: 'Domain not found' })

    return reply.send({
      domain: row,
      instructions: buildInstructions(row.domain, row.verification_token),
    })
  })

  // POST /api/v1/domains/:id/verify
  app.post<{ Params: { id: string }; Body: VerifyDomainInput }>(
    '/domains/:id/verify',
    async (req, reply) => {
      const method = req.body?.method ?? 'dns_txt'
      if (method !== 'dns_txt' && method !== 'well_known_file') {
        return reply
          .status(400)
          .send({ error: 'method must be "dns_txt" or "well_known_file"' })
      }

      const row = await queryOne<Domain>(
        'SELECT * FROM domains WHERE id = $1 AND user_id = $2',
        [req.params.id, req.user.sub],
      )
      if (!row) return reply.status(404).send({ error: 'Domain not found' })

      if (row.verification_status === 'verified') {
        return reply.send({ domain: row, already_verified: true })
      }

      const success =
        method === 'dns_txt'
          ? await verifyDnsTxt(row.domain, row.verification_token)
          : await verifyWellKnownFile(row.domain, row.verification_token)

      const updated = await transaction(async (client) => {
        const newStatus = success ? 'verified' : 'failed'
        const { rows } = await client.query<Domain>(
          `UPDATE domains
           SET verification_status = $1,
               verification_method = $2,
               verified_at = $3
           WHERE id = $4
           RETURNING *`,
          [newStatus, method, success ? new Date() : null, row.id],
        )

        const auditAction = success ? 'domain_verified' : 'domain_verification_failed'
        await client.query(
          `INSERT INTO audit_log (user_id, action, target, ip_address, metadata)
           VALUES ($1, $2, $3, $4::inet, $5)`,
          [
            req.user.sub,
            auditAction,
            row.domain,
            req.ip || null,
            JSON.stringify({ domain_id: row.id, method }),
          ],
        )

        return rows[0]
      })

      if (!success) {
        return reply.status(422).send({
          domain: updated,
          error: `Verification failed via ${method}. Ensure the record/file is published and try again.`,
          instructions: buildInstructions(row.domain, row.verification_token),
        })
      }

      return reply.send({ domain: updated })
    },
  )

  // DELETE /api/v1/domains/:id
  app.delete<{ Params: { id: string } }>('/domains/:id', async (req, reply) => {
    const row = await queryOne<Domain>(
      'SELECT id FROM domains WHERE id = $1 AND user_id = $2',
      [req.params.id, req.user.sub],
    )
    if (!row) return reply.status(404).send({ error: 'Domain not found' })

    await query('DELETE FROM domains WHERE id = $1', [req.params.id])
    return reply.status(204).send()
  })
}

File: apps/engine/app/routers/domains.py
import uuid
from datetime import datetime, timezone
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_session
from app.dependencies import get_current_user
from app.models.audit_log import AuditLog
from app.models.domain import Domain
from app.models.user import User
from app.schemas.domain import (
    CreateDomainRequest,
    CreateDomainResponse,
    DnsInstructions,
    DomainListResponse,
    DomainOut,
    DomainWithInstructions,
    DnsTxtInstructions,
    VerificationInstructions,
    VerifyDomainRequest,
    VerifyResponse,
    WellKnownFileInstructions,
    WellKnownInstructions,
)
from app.services.domain_verification import verify_dns_txt, verify_well_known_file

router = APIRouter(prefix="/domains", tags=["domains"])

AuthDep = Annotated[User, Depends(get_current_user)]
SessionDep = Annotated[AsyncSession, Depends(get_session)]


def _build_instructions(domain_name: str, token: str) -> VerificationInstructions:
    return VerificationInstructions(
        domain=domain_name,
        token=token,
        dns_txt=DnsTxtInstructions(
            record_name=f"_recon-verify.{domain_name}",
            record_value=token,
        ),
        well_known_file=WellKnownFileInstructions(
            url=f"https://{domain_name}/.well-known/recon-verification.txt",
            file_path=".well-known/recon-verification.txt",
            content=token,
        ),
    )


async def _get_owned_domain(
    domain_id: uuid.UUID,
    user: User,
    session: AsyncSession,
) -> Domain:
    domain = await session.get(Domain, domain_id)
    if domain is None or domain.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Domain not found")
    return domain


async def _write_audit(
    session: AsyncSession,
    user_id: uuid.UUID,
    action: str,
    target: str,
    request: Request,
    metadata: dict | None = None,
) -> None:
    log = AuditLog(
        user_id=user_id,
        action=action,
        target=target,
        ip_address=request.client.host if request.client else None,
        metadata_=metadata or {},
    )
    session.add(log)


# ── GET /domains ──────────────────────────────────────────────────────────────

@router.get("", response_model=DomainListResponse)
async def list_domains(
    current_user: AuthDep,
    session: SessionDep,
) -> DomainListResponse:
    rows = await session.scalars(
        select(Domain)
        .where(Domain.user_id == current_user.id)
        .order_by(Domain.created_at.desc())
    )
    return DomainListResponse(domains=[DomainOut.model_validate(d) for d in rows])


# ── POST /domains ─────────────────────────────────────────────────────────────

@router.post("", response_model=CreateDomainResponse, status_code=status.HTTP_201_CREATED)
async def create_domain(
    body: CreateDomainRequest,
    request: Request,
    current_user: AuthDep,
    session: SessionDep,
) -> CreateDomainResponse:
    token = f"recon-verify-{uuid.uuid4()}"
    domain = Domain(
        user_id=current_user.id,
        domain=body.domain,
        verification_token=token,
    )
    session.add(domain)
    try:
        await session.flush()
        await _write_audit(
            session, current_user.id, "domain_registered", body.domain, request,
            metadata={"domain_id": str(domain.id)},
        )
        await session.commit()
        await session.refresh(domain)
    except IntegrityError:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Domain already registered for your account",
        )

    return CreateDomainResponse(
        domain_id=domain.id,
        verification_token=token,
        dns_instructions=DnsInstructions(
            name=f"_recon-verify.{domain.domain}",
            value=token,
        ),
        well_known_instructions=WellKnownInstructions(
            url=f"https://{domain.domain}/.well-known/recon-verification.txt",
            value=token,
        ),
    )


# ── GET /domains/{id} ─────────────────────────────────────────────────────────

@router.get("/{domain_id}", response_model=DomainWithInstructions)
async def get_domain(
    domain_id: uuid.UUID,
    current_user: AuthDep,
    session: SessionDep,
) -> DomainWithInstructions:
    domain = await _get_owned_domain(domain_id, current_user, session)
    return DomainWithInstructions(
        domain=DomainOut.model_validate(domain),
        instructions=_build_instructions(domain.domain, domain.verification_token),
    )


# ── POST /domains/{id}/verify ─────────────────────────────────────────────────

@router.post("/{domain_id}/verify", response_model=VerifyResponse)
async def verify_domain(
    domain_id: uuid.UUID,
    body: VerifyDomainRequest,
    request: Request,
    current_user: AuthDep,
    session: SessionDep,
) -> VerifyResponse:
    domain = await _get_owned_domain(domain_id, current_user, session)

    if domain.verification_status == "verified":
        return VerifyResponse(
            domain=DomainOut.model_validate(domain),
            already_verified=True,
        )

    if body.method == "dns_txt":
        success = await verify_dns_txt(domain.domain, domain.verification_token)
    else:
        success = await verify_well_known_file(domain.domain, domain.verification_token)

    domain.verification_status = "verified" if success else "failed"
    domain.verification_method = body.method
    domain.verified_at = datetime.now(timezone.utc) if success else None

    audit_action = "domain_verified" if success else "domain_verification_failed"
    await _write_audit(
        session,
        current_user.id,
        audit_action,
        domain.domain,
        request,
        metadata={"domain_id": str(domain.id), "method": body.method},
    )
    await session.commit()
    await session.refresh(domain)

    if not success:
        return VerifyResponse(
            domain=DomainOut.model_validate(domain),
            error=(
                f"Verification failed via {body.method}. "
                "Ensure the record / file is published and try again."
            ),
            instructions=_build_instructions(domain.domain, domain.verification_token),
        )

    return VerifyResponse(domain=DomainOut.model_validate(domain))


# ── DELETE /domains/{id} ──────────────────────────────────────────────────────

@router.delete("/{domain_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_domain(
    domain_id: uuid.UUID,
    current_user: AuthDep,
    session: SessionDep,
) -> None:
    domain = await _get_owned_domain(domain_id, current_user, session)
    await session.delete(domain)
    await session.commit()
